files/private-chef-ctl-commands/key-control.rb

```diff
@@ -0,0 +1,255 @@
+        row[:public_key].each_line do |line|
+          puts "        #{line}"
+        end
+      end
+    end
+  end
+end
+
+class ArgParser
+  def find_arg_after_identifier(args, identifier)
+    result = nil
+    args.each_with_index do |arg, i|
+      if arg == identifier
+        result = args[i+1]
+        break
+      end
+    end
+    result
+  end
+end
+
+class AddKey
+  def check_org_args(cmd_args, usage_msg)
+    unless cmd_args.length == 2 || \
+      (cmd_args.length == 4 && (cmd_args[2] == "--expiration-date" || cmd_args[2] == "--key-name")) || \
+      (cmd_args.length == 6 && ((cmd_args[2] == "--expiration-date" && cmd_args[4] == "--key-name") || (cmd_args[4] == "--expiration-date" && cmd_args[2] == "--key-name")))
+      STDERR.puts usage_msg
+      STDERR.puts "Usage: Expiration date defaults to infinite. Pass an ISO 8601 fomatted string: YYYY-MM-DDTHH:MM:SS e.g. 2013-12-24T21:00:00"
+      STDERR.puts "Usage: Default name used is the fingerprint of the key passed."
+      exit 1
+    end
+  end
+
+  def check_client_args(cmd_args, usage_msg)
+    unless cmd_args.length == 3 || \
+      (cmd_args.length == 5 && (cmd_args[3] == "--expiration-date" || cmd_args[3] == "--key-name")) || \
+      (cmd_args.length == 7 && ((cmd_args[3] == "--expiration-date" && cmd_args[5] == "--key-name") || (cmd_args[5] == "--expiration-date" && cmd_args[3] == "--key-name")))
+      STDERR.puts usage_msg
+      STDERR.puts "Usage: Expiration date defaults to infinite. Pass an ISO 8601 fomatted string: YYYY-MM-DDTHH:MM:SS e.g. 2013-12-24T21:00:00"
+      STDERR.puts "Usage: Default name used is the fingerprint of the key passed."
+      exit 1
+    end
+  end
+
+  def add_key(cmd_args, id)
+    # TODO should keyname be optional and default the key fingerprint?
+
+    # this is guaranteed to work because it would have crashed above if not
+    # expiration_date and key_name will be nil if they weren't passed
+    arg_parser = ArgParser.new
+    expiration_date = arg_parser.find_arg_after_identifier(cmd_args, "--expiration-date")
+    key_name        = arg_parser.find_arg_after_identifier(cmd_args, "--key-name")
+    user = cmd_args[0]
+    key = File.read(cmd_args[1])
+
+    unless /-----BEGIN PUBLIC KEY-----/.match(key)
+      STDERR.puts "Error: Invalid public key passed. Key must begin with:"
+      STDERR.puts "Error: -----BEGIN PUBLIC KEY-----"
+      exit 1
+    end
+
+    # if a key_name was not specified, use the fingerprint of the key
+    unless key_name
+      begin
+        openssl_key_object = OpenSSL::PKey::RSA.new(key)
+        data_string = OpenSSL::ASN1::Sequence([
+                                               OpenSSL::ASN1::Integer.new(openssl_key_object.public_key.n),
+                                               OpenSSL::ASN1::Integer.new(openssl_key_object.public_key.e)
+                                              ])
+        key_name = OpenSSL::Digest::SHA1.hexdigest(data_string.to_der).scan(/../).join(':')
```

Comment:
Thanks for doing additional research on this.

Also, if they specify a key_id, then the fingerprint won't get stored anyway.